database-chassis python http-server implementation

anamehra opened the pull request:
Signed-off-by: anamehra [anamehra@cisco.com](mailto:anamehra@cisco.com)
```
  Python HTTP Server service.
  This service runs in database-chassis docker based on the configuration
  provided in device/<platform>/chassisdb.conf

  Configuration parameters provided by chassisdb.conf are as follows:

  start_http_server: if "yes" or "1", http-server will be started on this
      node
  chassis_db_address   : IP address for http-server. This is the same IP used
   for the chassis db server
  http_server_port : Port to bind to, default: 8000
  http_server_dir  : HTTP server home directory path, default: /var/www/

# Example of config:
'''
start_http_server=yes
chassis_db_address=127.0.0.10
http_server_dir=/var/www/tftp
'''
```
<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it
Enable http-server as database-chassis service on Chassis supervisor to enable LCs to download image and config files from Supervisor.

#### How I did it
Added python3 http-server based implementation in database-chassis as a docker service run by supervisord.

#### How to verify it
Add device/<platform>/chassisdb.conf with the following data:
start_http_server=yes
chassis_db_address=127.0.0.10
http_server_dir=/var/www/tftp

Boot with the image with the above data.

put an image file at /var/www/tftp/filename
check 'systemctl status http-service'
curl http://http-server:8000/filename -o out_file
compare file_out with /var/www/tftp/filename

#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [x] 202205
- [ ] 202211

#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

#### Ensure to add label/tag for the feature raised. example - PR#2174 under sonic-utilities repo. where, Generic Config and Update feature has been labelled as GCU.

#### Link to config_db schema for YANG module changes
<!--
Provide a link to config_db schema for the table for which YANG model
is defined
Link should point to correct section on https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/doc/Configuration.md
-->

#### A picture of a cute animal (not mandatory but encouraged)

anamehra committed:
database-chassis python http-server implementation

  Python HTTP Server service.
  This service runs in database-chassis docker based on the configuration
  provided in device/<platform>/chassisdb.conf

  Configuration parameters provided by chassisdb.conf are as follows:

  start_http_server: if "yes" or "1", http-server will be started on this
      node
  chassis_db_address   : IP address for http-server. This is the same IP used
   for the chassis db server
  http_server_port : Port to bind to, default: 8000
  http_server_dir  : HTTP server home directory path, default: /var/www/

Signed-off-by: anamehra <anamehra@cisco.com>

anamehra commented:
@abdosi , FYI-
Thanks

abdosi reviewed (dismissed):
LGTM

abdosi commented:
@arlakshm / @judyjoseph  please help review.

anamehra commented:
@judyjoseph , @arlakshm , please let me know if you have any comments. I resolved all comments from our last meeting discussion in this PR. moved the http-server in database-chassis docker for chassis.

judyjoseph commented on dockers/docker-database/http-server:
```diff
@@ -0,0 +1,120 @@
+            syslog.syslog(syslog.LOG_INFO, 'start_http_server is not set, exiting...')
+
+def main():
+    #import pdb; pdb.set_trace()
```
We could remove this comment.

anamehra committed:
Removed commented code

qiluo-msft commented on dockers/docker-database/supervisord.conf.j2:
```diff
@@ -56,3 +56,19 @@ stdout_logfile=syslog
+
+{% if INSTANCES %}
+{%     for redis_inst, redis_items in INSTANCES.items() %}
+{% if redis_inst == 'redis_chassis' %}
```
>if [](http://example.com/codeflow?start=3&length=2)

Need 2 level indentation

anamehra replied:
Thanks for the review @qiluo-msft !
I will fix this.

anamehra replied:
fixed.

qiluo-msft commented on dockers/docker-database/Dockerfile.j2:
```diff
@@ -44,5 +44,6 @@ COPY ["files/sysctl-net.conf", "/etc/sysctl.d/"]
 COPY ["critical_processes", "/etc/supervisor"]
 COPY ["files/update_chassisdb_config", "/usr/local/bin/"]
 COPY ["flush_unused_database", "/usr/local/bin/"]
+COPY ["http-server", "/usr/local/bin/"]
```
>http-server [](http://example.com/codeflow?start=7&length=11)

Why put this service in database container? I do not see any dependency

anamehra replied:
My initial PR was to run this as host side service as a feature but in the last review in Chassis group, it was decided to use it inside database-chassis for now as it's only for internal image/data hosting for chassis modules. It runs only on Supervisor on Chassis and binds to same internal ip as redis_chassis.server.

abdosi replied:
@qiluo-msft yes we had discussion on this and if we want to run on host there was comment from Guohan to add this as Feature in Feature Table. But that was little deviation from current model where each feature is like docker with we can enable/disable the feature by doing `container stop/start`.  So in chassis subgroup we discussed and thought since that will be some infrastructure change for this scenario to unblock chassis use case we thought to make it part of `database-chassis` docker which is specific to chassis use case. Also in future if this is needed in host we can revisit.

qiluo-msft commented on dockers/docker-database/http-server:
```diff
@@ -0,0 +1,119 @@
+# Example of config:
+'''
+start_http_server=yes
+chassis_db_address=127.0.0.10
```
>127.0.0.10 [](http://example.com/codeflow?start=19&length=10)

hard-coded IP?

anamehra replied:
This is an example of the config. The chassis_db_address is used for redis_chassis.server and is hardcoded as of today, provided by chassisd.conf file by vendor in <platform> dir.
This http-serevr use the same IP to bind to service requests on internal midplane network.

qiluo-msft commented on dockers/docker-database/supervisord.conf.j2:
```diff
@@ -56,3 +56,19 @@ stdout_logfile=syslog
+[program: http-server]
+command=python3 /usr/local/bin/http-server
+priority=3
+autostart=true
```
>true [](http://example.com/codeflow?start=10&length=4)

no way to disable it

qiluo-msft commented on dockers/docker-database/http-server:
```diff
@@ -0,0 +1,119 @@
+
+from pathlib import Path
+
+HTTP_DEFAULT_BIND_PORT = "8000"
```
>8000 [](http://example.com/codeflow?start=26&length=4)

Do we have rules to protect this service?

anamehra replied:
I will investigate more on this and respond/implement. Thanks

abdosi replied:
@qiluo-msft your point is to have Control plane acl (ip table rules)   to explicitly permit this ?

qiluo-msft replied:
yes, I believe you also want to limit the access, for example by some rules on 5-tuple

anamehra replied:
@abdosi , @qiluo-msft , this port is on the internal network. Do we have any rules on other internal ip/ports which I can refer to?

anamehra replied:
Hi @qiluo-msft , Abhishek and I had a discussion on this. The current Sonic implementation does not have any rule defined for the midplane subnet yet and we need to have a rule for all kinds of access in the midplane subnet. We will have to define platform subnet parameters that caclmgrd could use to define the rules for midplane access for all kinds of midplane traffic in the chassis. I will open a case and we can use a different PR to define the general subnet access rule. Let me know if that is fine.

abdosi replied:
@qiluo-msft on chassis setup we have additional interface `eth1-midplane` both on supervisor and LC for internal traffic (eg: redis and http)  . We will add iptable rule to allow all traffic on that interface. That way http traffic will be protected from outside network (`eth0`) and will get permit only if coming from internal network. I will create separate for that

abdosi replied:
Created PR : https://github.com/sonic-net/sonic-host-services/pull/42 . With this PR any port 8000 traffic that is not coming on eth1-midplane will get drop and thus protected from external network access.

abdosi commented:
@qiluo-msft can you please review this again.

anamehra committed:
Update supervisord.conf.j2

    If the start_http_server is not set, the process exit with status 0 but it
    exits before reaching RUNNING state which happens after running for more than a sec.
    In such a scenario, supervisord restarts the process and mark state as FATAL. For exit
    with status 0, we do not want to restart the process.

    Fixed indentation

anamehra committed:
Make /var/www accessible from database docker

anamehra committed:
Merge branch 'sonic-net:master' into anamehra/http_server_chassisdb

gechiang commented:
@anamehra , can you resolve the conflict since the changes has been a while and require rebase/merge to resolve conflicts.
We can then ask @qiluo-msft  to approve/merge.
Thanks!

abdosi commented:
> @anamehra , can you resolve the conflict since the changes has been a while and require rebase/merge to resolve conflicts. We can then ask @qiluo-msft to approve/merge. Thanks!

we don;t need this for 202205. This will gets merged to master directly.

anamehra committed:
Resolved conflict

Signed-off-by: anamehra <anamehra@cisco.com>

anamehra committed:
Merge branch 'master' into anamehra/http_server_chassisdb

anamehra commented:
> @anamehra , can you resolve the conflict since the changes has been a while and require rebase/merge to resolve conflicts. We can then ask @qiluo-msft to approve/merge. Thanks!

resolved. Thanks!

arlakshm commented:
Lower priority now. Will revisit in the future
